src/ChainlinkOracleImpl.sol

```diff
@@ -0,0 +1,308 @@
+/// @dev This contract uses token = address(0) to refer to ETH.
+contract ChainlinkOracleImpl is OracleImpl {
+    /// -----------------------------------------------------------------------
+    /// libraries
+    /// -----------------------------------------------------------------------
+
+    using AddressUtils for address;
+    using TokenUtils for address;
+    using ChainlinkPairDetails for mapping(address => mapping(address => PairDetail));
+
+    /// -----------------------------------------------------------------------
+    /// errors
+    /// -----------------------------------------------------------------------
+
+    error InvalidPair_FeedNotSet(QuotePair qp);
+    error StalePrice(AggregatorV3Interface feed, uint256 timestamp);
+    error NegativePrice(AggregatorV3Interface feed, int256 price);
+
+    /// -----------------------------------------------------------------------
+    /// structs
+    /// -----------------------------------------------------------------------
+
+    struct InitParams {
+        address owner;
+        bool paused;
+        uint32 defaultStaleAfter;
+        SetTokenOverrideParams[] tokenOverrides;
+        SetPairDetailParams[] pairDetails;
+    }
+
+    struct SetTokenOverrideParams {
+        address token;
+        address tokenOverride;
+    }
+
+    struct SetPairDetailParams {
+        QuotePair quotePair;
+        PairDetail pairDetail;
+    }
+
+    struct PairDetail {
+        AggregatorV3Interface feed;
+        bool inverted;
+        uint32 staleAfter;
+    }
+
+    /// -----------------------------------------------------------------------
+    /// events
+    /// -----------------------------------------------------------------------
+
+    event SetDefaultStaleAfter(uint32 defaultStaleAfter);
+    event SetTokenOverrides(SetTokenOverrideParams[] params);
+    event SetPairDetails(SetPairDetailParams[] params);
+
+    /// -----------------------------------------------------------------------
+    /// storage
+    /// -----------------------------------------------------------------------
+
+    /// -----------------------------------------------------------------------
+    /// storage - constants & immutables
+    /// -----------------------------------------------------------------------
+
+    address public immutable chainlinkOracleFactory;
+
+    /// -----------------------------------------------------------------------
+    /// storage - mutables
+    /// -----------------------------------------------------------------------
+
+    /// slot 0 - 7 byte free
+
+    /// OwnableImpl storage
+    /// address internal $owner;
+    /// 20 bytes
+
+    /// PausableImpl storage
+    /// bool internal $paused;
+    /// 1 byte
+
+    /// default time until CL updates become stale
+    /// 4 bytes
+    uint32 internal $defaultStaleAfter;
+
+    /// slot 1 - 0 bytes free
+
+    /// overrides for tokens
+    /// 32 bytes
+    mapping(address => address) internal $_tokenOverrides;
```

Comment:
what exactly is this for? you can just arbitrarily have any token overridden to use a different token?